src/python/pants/backend/jvm/tasks/jvm_compile/rsc/rsc_compile.py

```diff
@@ -607,17 +612,20 @@ def _jdk_libs_paths_and_digest(self, hermetic_dist):
   def _jdk_libs_abs(self, nonhermetic_dist):
     return nonhermetic_dist.find_libs(self._JDK_LIB_NAMES)
 
-  def _on_invalid_compile_dependency(self, dep, compile_target):
+  def _on_invalid_compile_dependency(self, dep, compile_target, contexts):
     """Decide whether to continue searching for invalid targets to use in the execution graph.
 
-    If a necessary dep is a Scala dep and the root is Java, continue to recurse because
-    otherwise we'll drop the path between Zinc compile of the Java target and a Zinc
-    compile of a transitive Scala dependency.
+    If a necessary dep is a Rsc compiled dep and the root is a Zinc one, continue to recurse because
+    otherwise we'll drop the path between Zinc compile of the Zinc target and a Zinc
+    compile of a transitive Rsc dependency.
 
-    This is only an issue for graphs like J -> S1 -> S2, where J is a Java target,
-    S1/2 are Scala targets and S2 must be on the classpath to compile J successfully.
+    This is only an issue for graphs like J -> S1 -> S2, where J is a Zinc target,
+    S1/2 are Rsc targets and S2 must be on the classpath to compile J successfully.
     """
-    if dep.has_sources('.scala') and compile_target.has_sources('.java'):
-      return True
-    else:
-      return False
+    return contexts[compile_target][0].workflow.resolve_for_enum_variant({
+      'zinc-only': lambda : contexts[dep][0].workflow.resolve_for_enum_variant({
+        'rsc-then-zinc': True,
+        'zinc-only': False
+      }),
+      'rsc-then-zinc': lambda : False
+    })()
```

Comment:
After #7269, this could be:
```python
    return contexts[compile_target][0].workflow.resolve_for_enum_variant({
      'zinc-only': lambda : contexts[dep][0].workflow == _JvmCompileWorkflow.rsc_then_zinc,
      'rsc-then-zinc': lambda: False,
    })()
```

Reply:
I'll convert `compile_contexts` into a datatype in followup PR.

Reply:
Cool.

This is the bit that I was referring to as a bug fix.
> Could you please expand on what this is referring to?
It refers to the fact that this conditional was trying to model different workflows, but when I adjusted how workflows were chosen, I hadn't updated this.

Reply:
Oh oops, thank you for elaborating!